src/database.rs

```diff
@@ -152,6 +154,230 @@ pub fn current_revision<Db: ?Sized + Database>(db: &Db) -> Revision {
     db.zalsa().current_revision()
 }
 
+impl dyn Database {
+    pub fn as_serialize(&mut self) -> impl serde::Serialize + '_ {
+        persistence::SerializeDatabase {
+            runtime: self.zalsa().runtime(),
+            ingredients: persistence::SerializeIngredients(self.zalsa()),
+        }
+    }
+
+    pub fn deserialize<'db, D>(&mut self, deserializer: D) -> Result<(), D::Error>
+    where
+        D: serde::Deserializer<'db>,
+    {
+        persistence::DeserializeDatabase(self.zalsa_mut()).deserialize(deserializer)
+    }
+}
+
+mod persistence {
+    use crate::plumbing::Ingredient;
+    use crate::zalsa::Zalsa;
+    use crate::Runtime;
+
+    use std::fmt;
+
+    use serde::de;
+    use serde::ser::SerializeMap;
+
+    #[derive(serde::Serialize)]
+    #[serde(rename = "Database")]
+    pub struct SerializeDatabase<'db> {
+        pub runtime: &'db Runtime,
+        pub ingredients: SerializeIngredients<'db>,
+    }
+
+    pub struct SerializeIngredients<'db>(pub &'db Zalsa);
+
+    impl serde::Serialize for SerializeIngredients<'_> {
+        fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+        where
+            S: serde::Serializer,
+        {
+            let SerializeIngredients(zalsa) = self;
+
+            let mut ingredients = zalsa
+                .ingredients()
+                .filter(|ingredient| ingredient.should_serialize(zalsa))
+                .collect::<Vec<_>>();
+
+            ingredients.sort_by(|a, b| {
+                // Ensure structs are serialized before tracked functions, as deserializing a
+                // memo requires its input struct to have been deserialized.
```

Comment:
It's fine deferring this to later but that will mean that we always serialize all tracked structs even if some query instances are never serialized (that created those tracked structs)